### ISSUE TITLE
ERR_UNKNOWN_FILE_EXTENSION since Node.js 18.19.0 (works fine with 18.18.2)

### ISSUE THREAD

**Issue by mlenkeit** [TypeStrong/ts-node#2094]
### Search Terms

- ERR_UNKNOWN_FILE_EXTENSION
- Node.js 18.19.0
- Node.js 18.18.2

### Expected Behavior

`npx ts-node <script.ts>` works on Node.js `18.19.0` just as it does with `18.18.2`.

### Actual Behavior

Since the upgrade to Node.js `18.19.0`, the call fails with `ERR_UNKNOWN_FILE_EXTENSION`. 

All other potentially related parameters remained unchanged (e.g. `package-lock.json`, `ts-config.json`, etc.).

We discovered this with GitHub workflow runs that run periodically (with `schedule`) at short intervals; only difference since things started failing is the Node.js version.

### Steps to reproduce the problem

- set up Node.js 18.19.2
- set `"type": "module"` in `package.json`
- set `"esm": true` in `tsconfig.json`
- create `main.ts` with something like: `console.log('Hello world')`
- run `npx ts-node main.ts`

### Minimal reproduction

see https://github.com/mlenkeit/ts-node-repro

there's also a GitHub workflow to run this repo against Node.js `18.19.0` and `18.18.2`:
https://github.com/mlenkeit/ts-node-repro/actions/runs/7057725777

### Specifications

* ts-node version: v10.9.1
* node version: v18.19.0
* TypeScript version: v5.2.2
* tsconfig.json, if you're using one:
  ```
  {
    "extends": "@tsconfig/node18/tsconfig.json",
    "ts-node": {
      "esm": true
    },
    "compilerOptions": {
      "allowSyntheticDefaultImports": true,
      "moduleResolution": "Node",
      "module": "ESNext",
      "noImplicitAny": false,
      "strictNullChecks": true,
      "sourceMap": true,
      "outDir": "./dist",
      "resolveJsonModule": false,
      "types": [
        "node"
      ]
    },
    "exclude": ["node_modules"]
  }
  ```
* package.json:
  ```
  {
    "name": "ts-node-repro",
    "private": true,
    "version": "1.0.0",
    "description": "",
    "main": "index.js",
    "type": "module",
    "dependencies": {
      "@tsconfig/node18": "18.2.2",
      "@types/node": "18.16.1",
      "ts-node": "10.9.1",
      "typescript": "5.2.2"
    }
  }
  ```
* Operating system and version: macOS Sonoma 14.1.1

**Comment by michael42** [comment]
Workaround: You can use `node --loader ts-node/esm file.ts` instead of `ts-node --esm file.ts` (which causes a `ExperimentalWarning: --experimental-loader may be removed in the future`). It would be great if `ts-node` provided an entrypoint that calls `register` from `node:module`, which seems to be Node's preferred way of registering loaders, now.

**Comment by jetwiwo** [comment]
Same issue in node `v20.10.0` (current LTS)

**Comment by jetwiwo** [comment]
> > Workaround: You can use `node --loader ts-node/esm file.ts` instead of `ts-node --esm file.ts` (which causes a `ExperimentalWarning: --experimental-loader may be removed in the future`). It would be great if `ts-node` provided an entrypoint that calls `register` from `node:module`, which seems to be Node's preferred way of registering loaders, now.
> 
> This does not seem to work for me: `node --loader ts-node/esm -r tsconfig-paths/register ./scripts/deploy.ts`
> 
> ```
> (node:129) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
> --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));'
> (Use `node --trace-warnings ...` to show where the warning was created)
> /builds/ci-vct-dev/aws-platform-ci/aws-cdk/scripts/deploy.ts:1
> import { AWS_DOMAIN } from '../shared-lib/src/environment/compile';
> ^^^^^^
> ```

Make sure you have `"type": "module" in package.json or rename the file to `file.mts`

**Comment by ajvincent** [comment]
See #1997 - this sounds like a duplicate of that one.

**Comment by mlenkeit** [comment]
> See #1997 - this sounds like a duplicate of that one.

I agree that #1997 and this one probably share the same root cause. 

Yet, I think they differ in their implication: If `ts-node` changed the way it works on a major version bump of Node.js, I'd say that's acceptable. But if `ts-node` suddenly stopped working on a minor version bump of Node.js, I'd say it's a bug.

**Comment by MythicManiac** [comment]
Got bit by this just a few minutes ago as well, would agree with not expecting breaking changes with minor version bumps.

**Comment by rreeves8** [comment]
Does anyone know which version of node it works on ? Currently have this issue with node 20, running ts-node --esm in a module

**Comment by rcollette** [comment]
I'm using tsx with AWS CDK until this is resolved.

In cdk.json, change the app property to:

```
  "app": "npx tsx bin/app.ts",
```

**Comment by mlenkeit** [comment]
> Does anyone know which version of node it works on ? Currently have this issue with node 20, running ts-node --esm in a module

@rreeves8 it works on Node.js 18.18.2. To my knowledge, any newer version runs into the problem described here (incl. Node.js 20). For following this kind of problem for Node.js 20, you may want to take a look at #1997 as well.

**Comment by kuncevic** [comment]
`node --loader ts-node/esm index.ts` works for me

**Comment by seemrcola** [comment]
> > > Workaround: You can use `node --loader ts-node/esm file.ts` instead of `ts-node --esm file.ts` (which causes a `ExperimentalWarning: --experimental-loader may be removed in the future`). It would be great if `ts-node` provided an entrypoint that calls `register` from `node:module`, which seems to be Node's preferred way of registering loaders, now.
> > 
> > 
> > This does not seem to work for me: `node --loader ts-node/esm -r tsconfig-paths/register ./scripts/deploy.ts`
> > ```
> > (node:129) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
> > --import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));'
> > (Use `node --trace-warnings ...` to show where the warning was created)
> > /builds/ci-vct-dev/aws-platform-ci/aws-cdk/scripts/deploy.ts:1
> > import { AWS_DOMAIN } from '../shared-lib/src/environment/compile';
> > ^^^^^^
> > ```
> 
> Make sure you have `"type": "module"` in package.json or rename the file to `file.mts`

thanks, it works for me

**Comment by mlenkeit** [comment]
FWIW, we have decided to switch to `tsx`, given that there hasn't been any reaction yet from the maintainers of `ts-node`.

In addition to replacing the dependency itself, we have removed the `ts-node` config from our `tsconfig.json` file and also uninstalled any explicit dependencies to `swc` (if present), as it seems to be the default in `tsx`.

We struggled a bit to make it work across Node.js versions, given that the `--loader` and `--import` options seem to be mutually exclusive, and ended up having dedicated scripts in our `package.json`.

**Comment by sovietspaceship** [comment]
Any update on this? This has rendered ts-node completely unusable and not fit for purpose. The `node --loader ts-node/esm file.ts` and similar workarounds do not always work, and break all existing projects that now have to temporarily implement the workaround while this gets fixed.

**Comment by qiulang** [comment]
The other workaround is to use [tsx](https://tsx.is/) instead. tsx also comes with watch mode, which is a bonus.

**Comment by andrewodri** [comment]
For anyone looking for a one-liner workaround in the future, you can implement @qiulang's suggestion by making the following change in your `cdk.json` file:

```diff
-   "app": "npx ts-node --prefer-ts-exts bin/project-cdk.ts",
+   "app": "npx tsx bin/project-cdk.ts",
```

Since both tsx is called via npx, you technically don't need to add it as a dependency... But it's probably a good idea to run an `npm i --save-dev tsx` to avoid any unexpected issues that may occur from tsx version changes in the future, especially if your code runs in CI.

**Comment by ctrlaltdylan** [comment]
@andrewodri 's answer works for me!

Just to make it super clear, you're not running `npx tsx bin/your-project-stack.ts` in the CLI. You're _updating_ your `cdk.json` file in the root directory of your project, _then_ running your normal `cdk` commands in the CLI.

More of a note to my future self for future projects because I'm bound to forget.

**Comment by Tuscan-blue** [comment]
@andrewodri Thank you! It helps me!

PS: Don't forget to set `"type": "module"` in the `package.json` file.
